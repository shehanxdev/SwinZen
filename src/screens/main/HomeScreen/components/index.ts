export * from './VideoUploadCard';
export * from './VideoThumbnailCard';
export * from './SubscribeCard';
export * from './SubscribeButton';
