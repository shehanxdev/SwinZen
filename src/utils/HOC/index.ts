export * from './linear-gradient-wrapper';
