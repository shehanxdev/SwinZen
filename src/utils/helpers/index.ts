export * from './common.helpers';
export * from './notification.helpers';
export * from './user.helpers';
