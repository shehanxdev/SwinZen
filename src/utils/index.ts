export * from './unit-tests';
export * from './validations';
