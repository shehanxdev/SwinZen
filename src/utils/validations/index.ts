export * from './auth';
export * from './info';
export * from './account';
