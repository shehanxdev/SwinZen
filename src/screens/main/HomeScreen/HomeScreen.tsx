import React, { useEffect, useMemo } from 'react';
import { View } from 'react-native';
import LinearGradient from 'react-native-linear-gradient';

import { Link, SwingZenLogoIcon, Text } from '@sz/components';
import { tw } from '@sz/config';
import { Color, Route, TextVariant } from '@sz/constants';
import { NavigationService, PermissionService } from '@sz/services';
import { useDispatch, useSelector } from '@sz/stores';

import { BaseMainScreen, MonthSelector } from '../components';
import { SubscribeCard, VideoUploadCard } from './components';
import { DataChart } from './components/DataChart/DataChart';
import { ProfileImageBanner } from './components/ProfileImageBanner';
import { UploadedVideoCountBanner } from './components/UploadedVideoCountBanner';

export function HomeScreen() {
  const dispatch = useDispatch();

  const loading = useSelector(state => state.loading.effects.userStore.fetchUserProfileData);
  const userProfileData = useSelector(state => state.userStore.profileData);
  const initialLogin = useSelector(state => state.persistentUserStore.loginState) === 'initial';

  useEffect(() => {
    //TODO::add proper error pop up to the user
    PermissionService.requestNotificationsPermission().catch(console.error);

    //fetch user profile data
    dispatch.userStore.fetchUserProfileData().catch(console.error);

    if (initialLogin) {
      setTimeout(() => NavigationService.navigate(Route.PricePlans), 0);
    } else {
      dispatch.userStore.getSubscription({}).catch(console.error);
    }
  }, []);

  const renderVideoAnalysisData = useMemo(() => {
    if (!userProfileData) return;

    //subscribed user
    if (userProfileData.isSubscribed) {
      //haven't update any videos
      if (userProfileData.chartData === null) return <VideoUploadCard />;
      return <DataChart {...userProfileData.chartData} />;
    }

    //user not subscribed
    return (
      <SubscribeCard features={['Analyzing your swing', 'Keep your performance noted', 'Unlimited video uploding']} />
    );
  }, [userProfileData]);

  const renderUploadedVideoCountData = useMemo(() => {
    if (!userProfileData) return;

    return (
      <View style={tw`mx-4 mt-5.5 mb-25`}>
        <UploadedVideoCountBanner
          videoCount={userProfileData.videoUploadData.videoUploads}
          swinZenUniVideoCount={userProfileData.videoUploadData.swingzenUniveristiy}
        />
      </View>
    );
  }, [userProfileData]);

  return (
    <BaseMainScreen>
      {loading ? (
        <Text variant={TextVariant.Heading3}>Loading</Text> //TODO::replace with proper loading indicator
      ) : (
        <>
          <View style={tw`mx-4 mt-5`}>
            <SwingZenLogoIcon width={70} height={34} />
            <View style={tw`mt-7`}>
              <ProfileImageBanner />
            </View>
            <View style={tw`mt-4.25`}>{renderVideoAnalysisData}</View>
          </View>
          {userProfileData?.isSubscribed ? (
            <LinearGradient
              colors={[Color.Neutral.Black, Color.Transparency.full, Color.Transparency.full]}
              locations={[0, 0.2206, 1]}
              style={tw`inset-0`}>
              <MonthSelector />
              <Text variant={TextVariant.Links}>
                {`Learn more about your progress. \nView your `} <Link text={'Instant Analysis Report.'} />
              </Text>
              {renderUploadedVideoCountData}
            </LinearGradient>
          ) : (
            renderUploadedVideoCountData
          )}
        </>
      )}
    </BaseMainScreen>
  );
}
