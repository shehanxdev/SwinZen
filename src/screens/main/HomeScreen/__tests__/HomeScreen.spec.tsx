import * as React from 'react';

import { renderWithProviders } from '@sz/utils';

import { HomeScreen } from './../HomeScreen';

describe('Home screen', () => {
  const getRenderedScreen = () => renderWithProviders(<HomeScreen />);

  it(`should render Home screen correctly`, async () => {
    const renderer = getRenderedScreen();
    const renderTree = renderer.toJSON();
    expect(renderTree).toMatchSnapshot();
  });
});
