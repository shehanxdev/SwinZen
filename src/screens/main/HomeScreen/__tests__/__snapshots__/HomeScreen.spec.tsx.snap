// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen should render Home screen correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <BVLinearGradient
    colors={
      [
        4279917603,
        4278984207,
        4278650631,
      ]
    }
    endPoint={
      {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
        "position": "relative",
      }
    }
  >
    <Image
      resizeMode="cover"
      source={
        {
          "testUri": "../../../src/assets/images/grassBackground.jpg",
        }
      }
      style={
        {
          "opacity": 0.1,
          "position": "absolute",
        }
      }
    />
    <RCTSafeAreaView
      style={
        {
          "height": "100%",
          "paddingTop": 50,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          [
            {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
            },
            {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <RCTScrollView
          contentContainerStyle={
            {
              "flexGrow": 1,
            }
          }
          keyboardShouldPersistTaps="handled"
        >
          <View>
            <View
              style={
                {
                  "marginBottom": "auto",
                  "marginLeft": "auto",
                  "marginRight": "auto",
                  "marginTop": "auto",
                }
              }
            >
              <View
                accessibilityRole="progressbar"
                accessibilityState={
                  {
                    "busy": true,
                  }
                }
                accessible={true}
                style={
                  [
                    {
                      "alignItems": "center",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "height": 48,
                      "opacity": 1,
                      "width": 48,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      {
                        "alignItems": "center",
                        "bottom": 0,
                        "justifyContent": "center",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        {
                          "height": 48,
                          "transform": [
                            {
                              "rotate": "45deg",
                            },
                          ],
                          "width": 48,
                        }
                      }
                    >
                      <View
                        collapsable={false}
                        style={
                          {
                            "height": 24,
                            "overflow": "hidden",
                            "width": 48,
                          }
                        }
                      >
                        <View
                          collapsable={false}
                          style={
                            {
                              "height": 48,
                              "transform": [
                                {
                                  "translateY": 0,
                                },
                                {
                                  "rotate": "-165deg",
                                },
                              ],
                              "width": 48,
                            }
                          }
                        >
                          <View
                            collapsable={false}
                            style={
                              {
                                "height": 24,
                                "overflow": "hidden",
                                "width": 48,
                              }
                            }
                          >
                            <View
                              collapsable={false}
                              style={
                                {
                                  "borderColor": "#7DC32C",
                                  "borderRadius": 24,
                                  "borderWidth": 4.8,
                                  "height": 48,
                                  "width": 48,
                                }
                              }
                            />
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                  <View
                    collapsable={false}
                    style={
                      {
                        "alignItems": "center",
                        "bottom": 0,
                        "justifyContent": "center",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        {
                          "height": 48,
                          "transform": [
                            {
                              "rotate": "45deg",
                            },
                          ],
                          "width": 48,
                        }
                      }
                    >
                      <View
                        collapsable={false}
                        style={
                          {
                            "height": 24,
                            "overflow": "hidden",
                            "top": 24,
                            "width": 48,
                          }
                        }
                      >
                        <View
                          collapsable={false}
                          style={
                            {
                              "height": 48,
                              "transform": [
                                {
                                  "translateY": -24,
                                },
                                {
                                  "rotate": "345deg",
                                },
                              ],
                              "width": 48,
                            }
                          }
                        >
                          <View
                            collapsable={false}
                            style={
                              {
                                "height": 24,
                                "overflow": "hidden",
                                "width": 48,
                              }
                            }
                          >
                            <View
                              collapsable={false}
                              style={
                                {
                                  "borderColor": "#7DC32C",
                                  "borderRadius": 24,
                                  "borderWidth": 4.8,
                                  "height": 48,
                                  "width": 48,
                                }
                              }
                            />
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </RCTSafeAreaView>
  </BVLinearGradient>
</View>
`;
